### PR TITLE
Jsonnet: Add ruler auto-scaling support

### PR DESCRIPTION
**Pull request by aknuds1**
#### What this PR does
Add support to Jsonnet for ruler auto-scaling, plus auto-scaling panels to Ruler dashboard.

I have tested the auto-scaling functionality internally, and it's equivalent to the prototype (i.e. it should be fine to incorporate into Mimir Jsonnet).

#### Checklist

- [ ] Tests updated
- [na] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by treid314 — APPROVED** [review]
LGTM! Thanks for adding the dashboard panel too!

**Review by pracucci — APPROVED** [review]
LGTM! I left a couple of suggestions for a follow up PR.

File: operations/mimir-mixin/dashboards/ruler.libsonnet [hunk 1]
```diff
@@ -101,6 +101,108 @@ local filename = 'mimir-ruler.json';
         $.latencyPanel('cortex_ingester_client_request_duration_seconds', '{%s, operation="/cortex.Ingester/QueryStream"}' % $.jobMatcher($._config.job_names.ruler))
       )
     )
+    .addRowIf(
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply operations/mimir-mixin/dashboards/ruler.libsonnet:104]: @pracucci thanks, I was wondering how we might refactor this code.

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,10 @@
 * [ENHANCEMENT] Alerts: Added `MimirAutoscalerKedaFailing` alert firing when a KEDA scaler is failing. #4045
 * [BUGFIX] Alerts: Fixed `MimirAutoscalerNotActive` to not fire if scaling metric does not exist, to avoid false positives on scaled objects with 0 min replicas. #4045
 
+### Jsonnet
+
+* [ENHANCEMENT] Add support for ruler auto-scaling. #4046
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply CHANGELOG.md:14]: @pracucci ah yes, good point.

File: operations/mimir-mixin/dashboards/ruler.libsonnet (hunk 1 above)
[the remark this answers is not in the document]
Comment (pracucci) [reply operations/mimir-mixin/dashboards/ruler.libsonnet:104]: > similar to kvStoreRow()

Does this answer your question or need more guidance?

[the remark this answers is not in the document]
Comment (aknuds1) [reply operations/mimir-mixin/dashboards/ruler.libsonnet:104]: @pracucci I will have to check if I can figure it out.

File: CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply CHANGELOG.md:15]: Done.

File: operations/mimir-mixin/dashboards/ruler.libsonnet (hunk 1 above)
[the remark this answers is not in the document]
Comment (aknuds1) [reply operations/mimir-mixin/dashboards/ruler.libsonnet:104]: @pracucci I could refactor into a function `cpuAndMemoryBasedAutoScalingRow`.

File: operations/mimir/autoscaling.libsonnet [hunk 3]
```diff
@@ -283,6 +295,55 @@
     if !$._config.autoscaling_distributor_enabled then {} else removeReplicasFromSpec
   ),
 
+  // Ruler
+
+  local newRulerScaledObject(name) = self.newScaledObject(
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply operations/mimir/autoscaling.libsonnet:300]: @treid314 did you work on something like this?

**Pull request merged**